File: Demos/04_ReverseLinkedList_Demo.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None

def ReverseLinkedList(head):
    node = head
    prevNode = None
    nextNode = None

    while node:
        nextNode = node.next
        node.next = prevNode
        prevNode = node
        node = nextNode

    return prevNode

def PythonListToLinkedList(pythonList):
    head = None
    tail = None

    for value in pythonList:
        node = Node(value)

        if head == None:
            head = node
            tail = head
            continue

        tail.next = node
        tail = node

    return head

def LinkedListToPythonList(head):
    pythonList = []
    while head:
        pythonList.append(head.val)
        head = head.next
    return pythonList

if __name__ == "__main__":
    nums = [i for i in range(10)]

    linkedList = PythonListToLinkedList(nums)
    print(LinkedListToPythonList(linkedList))

    linkedList = ReverseLinkedList(linkedList)
    print(LinkedListToPythonList(linkedList))

File: Demos/C++/04_ReverseLinkedList_demo.cpp
#include <iostream>

struct Node
{
    int value;
    Node* next;
    Node(int value) : value(value), next(nullptr) {}
};

void ReverseList(Node*& head)
{
    Node* node = head;
    Node* prevNode = nullptr;
    Node* nextNode = nullptr;

    while (node != nullptr)
    {
        nextNode = node->next;
        node->next = prevNode;
        prevNode = node;
        node = nextNode;
    }

    head = prevNode;
}

void PrintList(Node* head)
{
    Node* node = head;

    while (node != nullptr)
    {
        std::cout << node->value << " ";
        node = node->next;
    }
    std::cout << std::endl;
}

int main()
{
    // Build linked list with values ranging from 0 to 9
    Node* head = nullptr;
    Node* tail = nullptr;

    for (int i = 0; i < 10; i++)
    {
        Node* node = new Node(i);

        if (head == nullptr)
        {
            head = node;
            tail = node;
            continue;
        }

        tail->next = node;
        tail = node;
    }

    PrintList(head);
    ReverseList(head);
    PrintList(head);

    return 0;
}